lighthouse-core/audits/seo/robots-txt.js

```diff
@@ -0,0 +1,207 @@
+    let sitemapUrl;
+
+    try {
+      sitemapUrl = new URL(directiveValue);
+    } catch (e) {
+      throw new Error('Invalid sitemap URL');
+    }
+
+    if (!SITEMAP_VALID_PROTOCOLS.has(sitemapUrl.protocol)) {
+      throw new Error('Invalid sitemap URL protocol');
+    }
+  }
+
+  if (directiveName === DIRECTIVE_USER_AGENT && !directiveValue) {
+    throw new Error('No user-agent specified');
+  }
+
+  if (directiveName === DIRECTIVE_ALLOW || directiveName === DIRECTIVE_DISALLOW) {
+    if (directiveValue !== '' && directiveValue[0] !== '/' && directiveValue[0] !== '*') {
+      throw new Error('Pattern should either be empty, start with "/" or "*"');
+    }
+
+    const dollarIndex = directiveValue.indexOf('$');
+
+    if (dollarIndex !== -1 && dollarIndex !== directiveValue.length - 1) {
+      throw new Error('"$" should only be used at the end of the pattern');
+    }
+  }
+}
+
+/**
+ * @param {!string} line single line from a robots.txt file
+ * @throws will throw an exception if given line has errors
+ * @returns {!{directive: string, value: string}}
+ */
+function parseLine(line) {
+  const hashIndex = line.indexOf('#');
+
+  if (hashIndex !== -1) {
+    line = line.substr(0, hashIndex);
+  }
+
+  line = line.trim();
+
+  if (line.length === 0) {
+    return {};
+  }
+
+  const colonIndex = line.indexOf(':');
+
+  if (colonIndex === -1) {
+    throw new Error('Syntax not understood');
+  }
+
+  const directiveName = line.slice(0, colonIndex).trim().toLowerCase();
+  const directiveValue = line.slice(colonIndex + 1).trim();
+
+  verifyDirective(directiveName, directiveValue);
+
+  return {
+    directive: directiveName,
+    value: directiveValue,
+  };
+}
+
+function validateRobots(content) {
+  let inGroup = false;
+
+  return content
+    .split(/\r\n|\r|\n/)
+    .map((line, index) => {
+      let parsedLine;
+
+      try {
+        parsedLine = parseLine(line);
+      } catch (e) {
+        return {
+          index: index + 1,
+          line: line,
+          error: e.message,
+        };
+      }
+
+      // group-member records (allow, disallow) have to be precided with a start-of-group record (user-agent)
+      // see: https://developers.google.com/search/reference/robots_txt#grouping-of-records
+      if (parsedLine.directive === DIRECTIVE_USER_AGENT) {
+        inGroup = true;
+      } else if (!inGroup && DIRECTIVES_GROUP_MEMBERS.has(parsedLine.directive)) {
+        return {
+          index: index + 1,
+          line: line,
+          error: 'No user-agent specified',
+        };
+      }
+
+      return null;
+    })
+    .filter(error => error !== null);
+}
+
+class RobotsTxt extends Audit {
+  /**
+   * @return {!AuditMeta}
+   */
+  static get meta() {
+    return {
+      name: 'robots-txt',
+      description: 'robots.txt is valid',
+      failureDescription: 'robots.txt is not valid',
+      helpText: 'If your robots.txt file is malformed, crawlers may not be able to understand ' +
```

Comment:
do we have a link we can give them? perhaps your favorite validator/resource :)

Reply:
Since we are trying to stay crawler agnostic I think we shouldn't link to any of the official google/bing/yandex/yahoo docs here. Only independent resource that comes to mind is http://www.robotstxt.org/ but TBH it wasn't that useful to me. Proper docs for this audit are coming (#4355) to developers.google.com, so should wait for them.